### PR TITLE
ASoC: intel: sof_rt5682: Add support for tgl_rt1011_rt5682

### PR DESCRIPTION
**Pull request by brentlu**
Add support for two rt1011 speaker amplifier on SSP1 and ALC5682 on
SSP0 for TGL platform.

Signed-off-by: Brent Lu <brent.lu@intel.com>

### CODE REVIEW

File: sound/soc/intel/boards/Makefile
Comment (bardliao) [inline sound/soc/intel/boards/Makefile:21]: sof_realtek_common seems to contain rt1011 only. Should it be sof_i2s_rt1011 instead?

File: sound/soc/intel/boards/sof_realtek_common.c
Comment (bardliao) [inline sound/soc/intel/boards/sof_realtek_common.c:42]: What about RT1011_DEV2_NAME and RT1011_DEV3_NAME?

File: sound/soc/intel/boards/sof_realtek_common.c
Comment (bardliao) [inline sound/soc/intel/boards/sof_realtek_common.c:53]: There is no RT1011_DEV2_NAME and RT1011_DEV3_NAME here. Do we want to support 4 speakers in the driver?

File: sound/soc/intel/boards/sof_realtek_common.c
Comment (bardliao) [inline sound/soc/intel/boards/sof_realtek_common.c]: Will ``switch`` be better here? 
```
switch (i) {
        case 0: /* device 0 */
        tx_mask = 0x04;
        rx_mask = 0x01;
        ...
        break;
        case 1: /* device 1 */
        tx_mask = 0x08;
        rx_mask = 0x02;
        ...
        break;
        default:
        ...
        break;
}
ret = snd_soc_dai_set_tdm_slot(codec_dai, tx_mask,
                               rx_mask, 4, params_width(params));
if (ret < 0)
        break;
```

It can easily extend to supporting 4 speakers.

File: sound/soc/intel/boards/sof_realtek_common.c
Comment (bardliao) [inline sound/soc/intel/boards/sof_realtek_common.c:22]: You don't need to separate "Left Spk" and "Right Spk" unless you want to control them separately. See https://github.com/thesofproject/linux/pull/2585

**Review by bardliao — CHANGES_REQUESTED** [review]
@brentlu Thanks for the PR. It looks quite good. Few comments from me.

File: sound/soc/intel/boards/Makefile
Comment (brentlu) [reply sound/soc/intel/boards/Makefile:21]: I plan to move rt1015 part to this file in another patch when upstreaming to alsa-dev so I choose this name.

File: sound/soc/intel/boards/sof_realtek_common.c
Comment (brentlu) [reply sound/soc/intel/boards/sof_realtek_common.c:53]: Customer's board only supports two speakers, I can implement the code but not able to validate it.

Comment (brentlu) [reply sound/soc/intel/boards/sof_realtek_common.c:53]: BTW, do you know why i cannot create a pull request for sof repository? need to upload the topology file for this board.

File: sound/soc/intel/boards/Makefile
Comment (bardliao) [reply sound/soc/intel/boards/Makefile:21]: > I plan to move rt1015 part to this file in another patch when upstreaming to alsa-dev so I choose this name.

Sounds good.

File: sound/soc/intel/boards/sof_realtek_common.c
Comment (bardliao) [reply sound/soc/intel/boards/sof_realtek_common.c:53]: > Customer's board only supports two speakers, I can implement the code but not able to validate it.

Maybe add a comment to describe that this driver support two speaks for now.

Comment (bardliao) [reply sound/soc/intel/boards/sof_realtek_common.c:53]: > BTW, do you know why i cannot create a pull request for sof repository? need to upload the topology file for this board.

Please file an issue for it on SOF repository.

File: sound/soc/intel/boards/sof_realtek_common.c
Comment (RanderWang) [inline sound/soc/intel/boards/sof_realtek_common.c]: please add more info to dbg message like ("can't set to %d", srate)

File: sound/soc/intel/boards/sof_realtek_common.c
Comment (RanderWang) [inline sound/soc/intel/boards/sof_realtek_common.c]: add blank line

File: sound/soc/intel/boards/sof_realtek_common.c
Comment (lyakh) [inline sound/soc/intel/boards/sof_realtek_common.c:19]: `#include <sound/soc-dapm.h>`

File: sound/soc/intel/boards/sof_realtek_common.c
Comment (lyakh) [inline sound/soc/intel/boards/sof_realtek_common.c:64]: `#include <sound/pcm.h>`

File: sound/soc/intel/boards/sof_realtek_common.c
Comment (lyakh) [inline sound/soc/intel/boards/sof_realtek_common.c:65]: `#include <uapi/sound/asound.h>`

File: sound/soc/intel/boards/sof_realtek_common.c
Comment (lyakh) [inline sound/soc/intel/boards/sof_realtek_common.c:68]: `#include <sound/soc-dai.h>`

File: sound/soc/intel/boards/sof_realtek_common.c
Comment (lyakh) [inline sound/soc/intel/boards/sof_realtek_common.c]: `#include <linux/dev_printk.h>`

File: sound/soc/intel/boards/sof_realtek_common.c
Comment (lyakh) [reply sound/soc/intel/boards/sof_realtek_common.c]: or even make an array of mask structures, like
```
struct {
	unsigned int tx;
	unsigned int rx;
} mask[2] = {
	{.tx = 4, .rx = 1},
	{.tx = 8, .rx = 2},
};
...
ret = snd_soc_dai_set_tdm_slot(codec_dai, mask[i].tx,
                               mask[i].rx, 4, params_width(params));
```

File: sound/soc/intel/boards/sof_realtek_common.c
Comment (lyakh) [inline sound/soc/intel/boards/sof_realtek_common.c:129]: `#include <linux/kernel.h>`

**Review by lyakh — CHANGES_REQUESTED** [review]

File: sound/soc/intel/boards/sof_realtek_common.c
Comment (brentlu) [reply sound/soc/intel/boards/sof_realtek_common.c]: fixed

File: sound/soc/intel/boards/sof_realtek_common.c
Comment (brentlu) [reply sound/soc/intel/boards/sof_realtek_common.c:22]: maybe we can keep this flexibility to ucm?

File: sound/soc/intel/boards/sof_realtek_common.c
Comment (brentlu) [reply sound/soc/intel/boards/sof_realtek_common.c]: fixed

File: sound/soc/intel/boards/sof_realtek_common.c
Comment (brentlu) [reply sound/soc/intel/boards/sof_realtek_common.c]: fixed

File: sound/soc/intel/boards/sof_realtek_common.c
Comment (brentlu) [reply sound/soc/intel/boards/sof_realtek_common.c:19]: fixed

File: sound/soc/intel/boards/sof_realtek_common.c
Comment (brentlu) [reply sound/soc/intel/boards/sof_realtek_common.c:64]: fixed

File: sound/soc/intel/boards/sof_realtek_common.c
Comment (brentlu) [reply sound/soc/intel/boards/sof_realtek_common.c:65]: fixed

File: sound/soc/intel/boards/sof_realtek_common.c
Comment (brentlu) [reply sound/soc/intel/boards/sof_realtek_common.c:68]: fixed

File: sound/soc/intel/boards/sof_realtek_common.c
Comment (brentlu) [reply sound/soc/intel/boards/sof_realtek_common.c]: fixed

File: sound/soc/intel/boards/sof_realtek_common.c
Comment (brentlu) [reply sound/soc/intel/boards/sof_realtek_common.c:129]: fixed

File: sound/soc/intel/boards/sof_realtek_common.c
Comment (bardliao) [reply sound/soc/intel/boards/sof_realtek_common.c:22]: > maybe we can keep this flexibility to ucm?

Ok, let's keep the flexibility.

File: sound/soc/intel/boards/sof_realtek_common.c
Comment (bardliao) [inline sound/soc/intel/boards/sof_realtek_common.c]: We don't need to specify ``2`` here, ``[]`` is enough.

**Review by bardliao — COMMENTED** [review]
A minor comment. We usually don't specify array size.

File: sound/soc/intel/boards/sof_realtek_common.c
Comment (plbossart) [inline sound/soc/intel/boards/sof_realtek_common.c]: that's rather unusual to include this, can you revisit the includes?

**Review by plbossart — CHANGES_REQUESTED** [review]
minor change on includes, otherwise looks good. Thanks @brentlu

File: sound/soc/intel/boards/sof_realtek_common.c
Comment (brentlu) [reply sound/soc/intel/boards/sof_realtek_common.c]: removed

File: sound/soc/intel/boards/sof_realtek_common.c
Comment (brentlu) [reply sound/soc/intel/boards/sof_realtek_common.c]: done, i also use ARRAY_SIZE for the array size

**Comment by brentlu** [comment]
I'd plan to request ODM to validate this version of driver before sending to alsa-dev if there is no further change request.

**Comment by plbossart** [comment]
> I'd plan to request ODM to validate this version of driver before sending to alsa-dev if there is no further change request.

Or if you let us send the patch after it's merged in the SOF tree you'd get our Reviewed-by tags directly. We have a script that converts all approvals into tags.

File: sound/soc/intel/boards/sof_realtek_common.c
Comment (plbossart) [inline sound/soc/intel/boards/sof_realtek_common.c]: can this be static const? please make sure you compile with
make W=1 sound/soc/intel/boards/
make C=2 sound/soc/intel/boards/

File: sound/soc/intel/common/soc-acpi-intel-tgl-match.c
Comment (plbossart) [inline sound/soc/intel/common/soc-acpi-intel-tgl-match.c:317]: wondering if these tables can be const? I know all other tables are not, but I see no reason why.

**Review by plbossart — CHANGES_REQUESTED** [review]
need to check for static const declarations to avoid sparse warnings.

File: sound/soc/intel/boards/sof_realtek_common.c
Comment (brentlu) [reply sound/soc/intel/boards/sof_realtek_common.c]: W=1 is ok but C=2 fails to compile...

brentlu@brentlu-brix:/workspace/sof/linux$ make C=2 sound/soc/intel/boards/
  CHECK   scripts/mod/empty.c
/bin/sh: 1: sparse: not found
scripts/Makefile.build:283: recipe for target 'scripts/mod/empty.o' failed
make[1]: *** [scripts/mod/empty.o] Error 127
Makefile:1199: recipe for target 'prepare0' failed
make: *** [prepare0] Error 2

**Comment by brentlu** [comment]
> > I'd plan to request ODM to validate this version of driver before sending to alsa-dev if there is no further change request.
> 
> Or if you let us send the patch after it's merged in the SOF tree you'd get our Reviewed-by tags directly. We have a script that converts all approvals into tags.

Sure. Please ignore the patch sent to patchwork already...

File: sound/soc/intel/boards/sof_realtek_common.c
Comment (paulstelian97) [reply sound/soc/intel/boards/sof_realtek_common.c]: sparse is a tool you install, seems from your error messages that you don't have it.

File: sound/soc/intel/boards/sof_realtek_common.c
Comment (brentlu) [reply sound/soc/intel/boards/sof_realtek_common.c:42]: not use since we only support 2-amp configuration

File: sound/soc/intel/boards/sof_realtek_common.c
Comment (brentlu) [reply sound/soc/intel/boards/sof_realtek_common.c:53]: Add comment in line 16

File: sound/soc/intel/boards/sof_realtek_common.c
Comment (lyakh) [reply sound/soc/intel/boards/sof_realtek_common.c]: @plbossart it is true, that dev_printk.h isn't included directly very often, however, IMHO you need to include a header for `dev_err()` and friends. If not dev_printk.h then at least device.h which is the only header, that includes it.

Comment (brentlu) [reply sound/soc/intel/boards/sof_realtek_common.c]: will include linux/device.h

File: sound/soc/intel/common/soc-acpi-intel-tgl-match.c
Comment (brentlu) [reply sound/soc/intel/common/soc-acpi-intel-tgl-match.c:317]: i think it's ok to make it const since .quirk_data is only accessed by snd_soc_acpi_codec_list() without modifying it.

File: sound/soc/intel/boards/sof_realtek_common.c
Comment (brentlu) [reply sound/soc/intel/boards/sof_realtek_common.c]: Seems it crashes with or without my patch...

brentlu@brentlu-brix:/workspace/sof/linux$ make C=2 sound/soc/intel/boards/
  CHECK   scripts/mod/empty.c
<...>
./include/linux/restart_block.h:30:29: error: undefined identifier '__user'
./include/linux/restart_block.h:30:29: error: too many errors
Segmentation fault (core dumped)
scripts/Makefile.build:283: recipe for target 'sound/soc/intel/boards/sof_rt5682.o' failed
make[4]: *** [sound/soc/intel/boards/sof_rt5682.o] Error 139
scripts/Makefile.build:500: recipe for target 'sound/soc/intel/boards' failed
make[3]: *** [sound/soc/intel/boards] Error 2
scripts/Makefile.build:500: recipe for target 'sound/soc/intel' failed
make[2]: *** [sound/soc/intel] Error 2
scripts/Makefile.build:500: recipe for target 'sound/soc' failed
make[1]: *** [sound/soc] Error 2
Makefile:1799: recipe for target 'sound' failed
make: *** [sound] Error 2

File: sound/soc/intel/boards/sof_realtek_common.c
Comment (lyakh) [inline sound/soc/intel/boards/sof_realtek_common.c:56]: I would just place it in the function as a normal local variable, don't see why it needs a file scope?

File: sound/soc/intel/common/soc-acpi-intel-tgl-match.c
Comment (lyakh) [reply sound/soc/intel/common/soc-acpi-intel-tgl-match.c:317]: the important point here is, that `.quirk_data` is declared `const` in `struct snd_soc_acpi_mach`:
```
struct snd_soc_acpi_mach {
	...
	const void *quirk_data;
```
so, yes, it's good.

**Review by lyakh — APPROVED** [review]
would be good to have the `mask` array moved into a function, but it doesn't seem significant enough to ask for a new revision to me

**Review by RanderWang — APPROVED** [review]

**Comment by plbossart** [comment]
@brentlu please close this PR if the code is already upstream, thanks!

**Comment by plbossart** [comment]
closing since there was no update or information

**Pull request closed without merging**
